x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java

```diff
@@ -404,17 +413,94 @@ protected boolean initialRun() {
+        }
+
+        refreshDestinationIndex(new RefreshRequest(transformConfig.getDestination().getIndex()), ActionListener.wrap(response -> {
+            if (response.getFailedShards() > 0) {
+                logger.warn(
+                    "[{}] failed to refresh transform destination index, not all data might be available after checkpoint.",
+                    getJobId()
+                );
+            }
+            // delete data defined by retention policy
+            if (transformConfig.getRetentionPolicyConfig() != null) {
+                executeRetentionPolicy(listener);
+            } else {
+                finalizeCheckpoint(listener);
+            }
+        }, listener::onFailure));
+    }
 
+    private void executeRetentionPolicy(ActionListener<Void> listener) {
+        DeleteByQueryRequest deleteByQuery = RetentionPolicyToDeleteByQueryRequestConverter.buildDeleteByQueryRequest(
+            transformConfig.getRetentionPolicyConfig(),
+            transformConfig.getSettings(),
+            transformConfig.getDestination(),
+            nextCheckpoint
+        );
+
+        if (deleteByQuery == null) {
+            finalizeCheckpoint(listener);
+            return;
+        }
+
+        logger.debug(
+            new ParameterizedMessage(
+                "[{}] Run delete based on retention policy using dbq [{}] with query: [{}]",
+                getJobId(),
+                deleteByQuery,
+                deleteByQuery.getSearchRequest()
+            )
+        );
+        getStats().markStartDelete();
+
+        doDeleteByQuery(deleteByQuery, ActionListener.wrap(bulkByScrollResponse -> {
+            logger.trace(new ParameterizedMessage("[{}] dbq response: [{}]", getJobId(), bulkByScrollResponse));
+            getStats().markEndDelete();
+            if (bulkByScrollResponse.getDeleted() > 0) {
+                getStats().incrementNumDeletedDocuments(bulkByScrollResponse.getDeleted());
+                logger.debug("[{}] deleted [{}] documents as part of the retention policy.", getJobId(), bulkByScrollResponse.getDeleted());
+            }
+
+            // this should not happen as part of checkpointing
+            if (bulkByScrollResponse.getVersionConflicts() > 0) {
+                logger.warn(
+                    new ParameterizedMessage(
+                        "[{}] found [{}] version conflicts when deleting documents as part of the retention policy. "
+                            + "Retry at the next checkpoint.",
+                        getJobId(),
+                        bulkByScrollResponse.getDeleted()
+                    )
+                );
+            }
+            // paranoia: we are not expecting dbq to fail for other reasons
+            if (bulkByScrollResponse.getBulkFailures().size() > 0 || bulkByScrollResponse.getSearchFailures().size() > 0) {
+                assert false : "delete by query failed unexpectedly" + bulkByScrollResponse;
+                logger.warn(
```

Comment:
I added this case, but actually think I will change it to call `onFailure`, it would be inconsistent to search and indexing errors, I mean that's why we have retry-logic.